PULL_REQUEST_TEMPLATE.md

```diff
@@ -1,12 +1,42 @@
-<!-- Your title above should be a short description of what
-was changed. Do not include the issue number in the title. -->
+<!--
+Thank you so much for your PR!  To help us review your contribution, please
+consider the following points:
+
+- Your title above should be a short description of what was changed. Do not
+  include the issue number in the title.
+
+- Development information is available at
+  https://github.com/sympy/sympy/wiki/Development-workflow.
+
+- Do not create the PR out of master, but out of a separate branch.
+
+- Try to keep the number of commits to a minimum. Use rebase and
+  force pushing in git to avoid adding unnecessary commits.
+
+- If this is your first commit, see:
+  https://github.com/sympy/sympy/wiki/Development-workflow#add-your-name-and-email-address-to-the-mailmap-file
+
+- If you are contributing fixes to docstrings, please pay attention to
+  https://docs.sympy.org/dev/guides/contributing/documentation-style-guide.html.
+  In particular, note the difference between using single backquotes,
+  double backquotes, and asterisks in the markup.
+
+- Please include tests of any new feature or bug fix. The code in the tests
+  should not execute correctly in a version of SymPy without your code.
+
+Note that all reviewing is done an a volunteer basis, so it can sometimes take
+time before you get feedback. If you have not heard back in two weeks, feel
+free to ping @sympy/developers-with-push-access-to-everything
+Do not ping individual persons to get a review, nor request reviews.
```

Comment:
I'm not sure why we'd want this as a rule. Why can't people ping and request reviews?

Reply:
It would be nice to setup more teams in the sympy org. We could create teams for packages in sympy and then people could request reviews or ping that team to get their attention.

Reply:
Well, there is a tendency that some typically newer, contributors ping a reviewer they got help from for every PR they make and even ping in for other PRs. Here is a random, but typical example: https://github.com/sympy/sympy/pull/22806

I can for sure be pinged in for reviews where I have something to contribute (for the above example, not much), but it is a bit annoying to be pinged in arbitrarily. Also, my feeling is that other reviewers are reluctant to "take over" a review, if a reviewer has been pinged in. (Or it is simply that not enough of the 50 people with push access are really using it? From that perspective, I believe that it is better that "all" potential reviewers are pinged than that the most active ones gets even more requests, which is basically what is happening now.)

Reply:
Regarding teams: sure. I was about to add that (and hadn't seen that comment when I starting replying). That would for sure make sense. (Also, the name @sympy/developers-with-push-access-to-everything is not really convenient. Yes, I deliberately left it there so that all members have a chance to say something, like volunteering for a team.)

Reply:
There is far too much pinging. Also to be clear @oscargus is right that I am personally less likely to look at a PR if someone else has been pinged so naming someone in particular is only worthwhile if there is a particular reason why that person might want to look at it. Many new contributors are unable to judge that distinction. Saying clearly that they can ping a group rather than a particular person is helpful.

Also I think that giving an explicit timeframe is good to set some level of expectation.

Reply:
> Well, there is a tendency that some typically newer, contributors ping a reviewer they got help from for every PR they make and even ping in for other PRs.

So we should say something like "Only ping individual developers if they are engaged with you in that specific work" or something like that. We can encourage to only ping teams.

Reply:
Also, I think the Github "Request Review" feature could be helpful for us if we use it and use it more consistently. We could encourage it's use. But it would be better to request a review from a team. For example, maybe there is a team that cares about maintaining solvers. Also if we use this https://github.com/sympy/sympy/blob/master/CODEOWNERS, then those people will automatically get a review request for certain code.

Reply:
I think we should set teams as code owners. I assume that you gets some notification (review request(?)) for physics, right? In that way, there is no need to ping. Unless it has taken a long time without any feedback.

Reply:
(Again, I didn't see your follow-up before starting typing. Hence, basically typing the same thing...)

I think the way to go forward is that teams are code owners (if that it possible) and then setup teams with people that care about and/or have knowledge about that part of the code base.

(I find that often to be a problem, I many (most) parts of the code base I can only review the code style etc, but not really the mathematical aspects or APIs or anything like that. So even if they follow my review comments, I will not merge it as I do not know if it the right/wanted way to solve it.)

Reply:
> I think the way to go forward is that teams are code owners (if that it possible) and then setup teams with people that care about and/or have knowledge about that part of the code base.

I am all for this idea. We just need to make some teams and then add some documentation explaining that you should request to be part of these teams if you'd like to be automatically notified for review requests. Being part of the team would also mean that you'll get notified when people ping the team.

Reply:
I completely agree to what @moorepants is saying . I want to put forward a non member though active/daily contributor's perspective . This is what I observed since I started contributing some time back when I was completely new to open source and sympy. (Won't be tagging/mentioning using **@** here as it would be lot of unnecessary tagging)

Though oscargus mentions that following the matplotlib module would be a good way to go here but I would like to address some points here .

- I see that the core members are more inclined/comfortable in particular domains . I clearly see this to be a good thing as obviously a member with more knowledge and comfort regarding a domain both mathematically and code vise should be responsible for merging/reviewing most Prs in that domain . Like from my somewhat experience I can see
1) moorepants  reviews `physics` mainly 
2) oscarbenjamin reviews `solvers /odes `
and I could go on .......

Hence pinging @sympy/developrs won't be of too much help , obviously I won't expect a member to help me with somewhat he doesn't know (like maybe pinging moorepants for geometry or oscarbenjamin for phsyics)

- Many times a core member  might review to the fullest but might shy away from merging or guiding further due to various reasons . You can easily find the last comment on many prs by new contributors and others to be something like `I have made the changes , does this look good` or `what would you like to me address as I have made requested changes` and the pr doesn't receive traction from this point onwards!

I mainly contribute to the `series/limits/functions/concrete` module by priority and have been lucky enough to receive reviews by all members but still the merge button for my prs regarding these modules have mostly been pressed by jksoum except once obviously because he deals with that module. The best pr from my side yet (I feel so ) (`22806`) had received great reviews and I feel is more than ready to go but still the original reviewer hasn't taken a decision yet regarding what has to be done and I agree this also led to some random pings from my side( I was too excited to get that pr in : ) ... hope jksoum sees the pr soon )

This confuses new/non core contributors more than anything I guess as there is no clear way to go here . Hence this leads to random pinging as pointed above and even after pinging the original reviewer (who just turned up to give general/basic reviews and might not be the best person or the responsible one for that module) might not reciprocate/respond as he might not be clear how to go about here . 

- I had tried out oscarbenjamin 's idea of trying to send a mail to the mailing list asking for reviews attaching all my prs which had recieved reviews a month back or two but that didn't help much as I had expected . Because as I said above we should be expecting reviews/merging by people who are inclined/most comfortable in a particular domains and not by all members .

------------------------------------------------------------------------------------------------------
Coming to what I feel is best here.....  all these points above call for **Teams** !!!

1) There should be an instruction in the pr template to ping @sympy/module eg @sympy/series for reviews if there hasn't been any reply for some time . This team could maybe be formed of members most comfortable students and contributors/students who did good work in that module ( previous gsocers for that module i guess) 

2) Don't get me wrong , anyone and everyone is free to review anything and everything but the pinging part ..... , this shouldn't be done for an individual members or all developers but a team of people like sympy/series . 

3) The minimum time for waiting for review will obviously be decided by you'll so can't may much on this .

4) A wiki page could be present showing github account or members under a team , so that even under worst circumstances ... if a contributor wants to tag an individual to have his views it's not from the wrong team , which is happening quite often at present .

I hope you'll like some of the thoughts here !

Reply:
I like what you say @anutosh491, but for this point:

> should be responsible for merging/reviewing most Prs in that domain

We can't make anyone responsible for merging, as we are all volunteers and for whatever reason may not be able to review or merge. Whatever we instate has to accommodate a flexible and optional participation mode.

Reply:
Yeah correct correct.. sorry for that, didn't really mean to put it as the way it came out .I didn't want to propose any strict rules or anything on what I wrote above ,just thought of conveying what a non member/new contributor might be thinking here ! 

Glad I could convey what I felt is the best way out here.

Reply:
On the other hand we can say that someone submitting a PR is responsible for making it as good as possible and also as easy to review as possible so before they consider pinging anyone:

1. Make sure that all previous comments have been properly responded to including checking whether the same comments apply to other parts of the PR and adding tests for any examples that come up in review/discussion.
2. Make sure that they have followed the various steps like adding tests, updating docs, release note etc.
3. Improve the title and description in the PR of what the changes are to make sure that it is easy for a reviewer to understand the purpose of the changes.
4. Consider whether the code can be simplified or improved.
5. Try to think of different possible inputs that might or might not fail under the proposed changes rather than waiting for a reviewer to come up with counter examples.
6. Consider performance: what happens if the same code is used with larger nontrivial examples? Is it faster or slower than master?
7. Do more research to consider whether there are better approaches than the one taken in the PR.

Although there are parts of the codebase that I am less familiar with that doesn't mean I won't review or merge a PR there if it is easy to understand what the problem is, what is being changed and what other impacts it might have. It is usually possible for the person submitting the PR to make it easier for me to understand those things if they take the time to explain it better (and if the code is well written).

Reply:
I agree with Jason here. It's true that some core people get pinged all the time (honestly I never even notice when I am pinged because I just get an email for every issue anyway). But there are some people who might not be subscribed to every issue that *do* want to be pinged when something relevant to them is changed. The text here seems to discourage that. 

Using teams is a good idea. That's basically the same thing as pinging individual people except you don't have to know about how to ping, you can just ping the relevant team. 

The issue in question is a separate issue, which is *over pinging*. You definitely shouldn't ping someone more than once unless there are new changes to look at.

Reply:
> On the other hand we can say that someone submitting a PR is responsible for making it as good as possible and also as easy to review as possible so before they consider pinging anyone:

@oscarbenjamin I like these points. I could be good to add a version to the documentation to give guidelines on what a good PR looks like and how to make it easy to review.

Reply:
> The issue in question is a separate issue, which is over pinging. You definitely shouldn't ping someone more than once unless there are new changes to look at.

I'd argue that there isn't an issue. In real life if I want to walk to my colleagues office every 10 minutes to ask him if he's made progress on my request, he'll eventually get annoyed and tell me to stop. Pinging someone on github is no different. If some people over ping, just politely ask them to not do so. In the introduction to contributing, I linked to http://shakthimaan.com/downloads/book/chapter1.pdf, because people also write emails to our list that don't fit the cultural norms. That link will not stop it from happening, but it may help reduce it some. I think we just have to be ok with these kinds of emails and overpinging if we want to be welcoming to newcomers. Just have your standard polite response. We will have to do this over and over and over. It is just the way it is. I teach at university and every single year I have to teach the new bunch the cultural norms. I just accept that it is a never ending process, because it is.

Reply:
> if you'd like to be automatically notified for review requests

I would not expect anyone to sign up under those circumstances... I'd rather say that being part of a team is more or less mandatory to obtain push rights.

Reply:
> You definitely shouldn't ping someone more than once unless there are new changes to look at.

There is no need to ping people then either as one will get notifications anyway (unless one has unsubscribed).